bdb/phys.c

```diff
@@ -255,18 +255,63 @@ int phys_dta_add(bdb_state_type *bdb_state, tran_type *logical_tran,
     int max_poll = bdb_state->attr->pagedeadlock_maxpoll;
     tran_type *physical_tran = NULL;
 
+    DBT *addlkptr = NULL;
+    DB_LOCK *addrowlk = NULL;
+    DB_LSN last_regop_lsn = logical_tran->last_regop_lsn;
+
     /* Start transaction */
-    rc = get_physical_transaction(bdb_state, logical_tran, &physical_tran);
-    if (rc)
+    rc = get_physical_transaction(bdb_state, logical_tran, &physical_tran, 0);
+    if (rc) {
+        logmsg(LOGMSG_ERROR, "%s failed get physical tran rc %d\n", __func__,
+               rc);
         goto done;
+    }
+
+    if (dtafile == 0 && !logical_tran->single_physical_transaction) {
+        int did_commit = 0;
+        rc = tran_allocate_rlptr(logical_tran, &addlkptr, &addrowlk);
+        if (rc) {
+            logmsg(LOGMSG_ERROR, "%s failed to allocate rlptr rc %d\n",
+                   __func__, rc);
+            goto done;
+        }
+
+        if (log_compare(&last_regop_lsn, &logical_tran->last_regop_lsn))
+            did_commit = 1;
+
+        rc = bdb_lock_row_write_getlock(bdb_state, logical_tran, -1, genid,
```

Comment:
I was convinced that we didn't need to get a writelock for dta file inserts .. my thinking was that given that genids are unique, it's impossible to have collisions here.